storage/bbolt.go

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (C) 2022 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/go-stoabs/bbolt"
+	"github.com/nuts-foundation/nuts-node/storage/log"
+	bboltLib "go.etcd.io/bbolt"
+	"os"
+	"path"
+	"sync"
+	"time"
+)
+
+const fileMode = 0640
+
+type bboltDatabase struct {
+	datadir         string
+	config          BBoltConfig
+	ctx             context.Context
+	cancel          context.CancelFunc
+	shutdownWatcher *sync.WaitGroup
+}
+
+// BBoltConfig specifies config for BBolt databases.
+type BBoltConfig struct {
+	// Backup specifies backup config for the database.
+	Backup BBoltBackupConfig `koanf:"backup"`
+}
+
+// BBoltBackupConfig specifies config for BBolt database backups.
+type BBoltBackupConfig struct {
+	// Directory specifies the directory in which the BBolt backup should be written.
+	Directory string `koanf:"directory"`
+	// Interval specifies the time between backups.
+	Interval time.Duration `koanf:"interval"`
+}
+
+func (b BBoltBackupConfig) Enabled() bool {
+	return b.Interval > 0 && len(b.Directory) > 0
+}
+
+func createBBoltDatabase(datadir string, config BBoltConfig) (*bboltDatabase, error) {
+	result := bboltDatabase{
+		datadir:         datadir,
+		config:          config,
+		shutdownWatcher: &sync.WaitGroup{},
+	}
+	// Create context for initiating shutdown
+	result.ctx, result.cancel = context.WithCancel(context.Background())
+	return &result, nil
+}
+
+func (b bboltDatabase) createStore(moduleName string, storeName string) (stoabs.KVStore, error) {
+	databasePath := path.Join(b.datadir, b.getRelativeStorePath(moduleName, storeName))
+	store, err := bbolt.CreateBBoltStore(databasePath)
+	if store != nil {
+		b.startBackup(moduleName, storeName, store)
+	}
+	return store, err
+}
+
+func (b bboltDatabase) getClass() Class {
+	return VolatileStorageClass
```

Comment:
If it has a backup running, then it's not volatile anymore?

Reply:
Probably? But if you configure a backup for BBolt, but configure a Redis server as well I'd say we'd give precedence to Redis for storing non-volatile data?

I don't feel like implementing it at this moment, because we don't support per-store database binding (e.g. DAG in Redis, network connections in BBolt) and backup is currently all or nothing (so having the separate classes won't change anything).

Change when we implemented Redis.